### DISCUSSION TITLE
Record derived metric, then drop the original data?

### DISCUSSION THREAD

**Discussion by dominikwerder** [prometheus/prometheus#13063] [Q&A]
Via recording rules I record a few derived metrics from a large set of scraped metrics.

Is it somehow possible to then drop the original scraped metrics to save space?

I couldn't find a way, so I currently use a 2-stage setup with federation where the 1st stage is very limited in disk space. There must be some easier solution to this but I couldn't figure it out.

**Comment by hamedrabah** [comment]
Recording rules cannot replace the source samples inside the same Prometheus. The rule evaluates against data that has already been ingested into the TSDB, then appends its derived series. Metric relabeling can drop samples before storage, but if it drops the inputs, the recording rule has nothing to evaluate.

Deleting the source series afterward through the admin API is also not a continuous solution. It races with new ingestion, creates tombstones/compaction work, and removes the data needed by later rule evaluations.

Your two-stage design is the normal way to create this boundary: keep the first Prometheus on short retention, evaluate the recording rules there, and expose or remote-write only the derived series to a long-retention system. Size the first stage for at least the longest rule lookback plus restart/replay margin.

If the transformation can happen before PromQL, an OpenTelemetry Collector or another stream processor could emit only the reduced metrics. For arbitrary PromQL recording rules, though, a short-retention first stage is the clean architecture rather than a workaround.
